core/arch/aarchxx/mangle.c

```diff
@@ -3043,7 +3043,13 @@ create_ldax_from_stex(dcontext_t *dcontext, instr_t *strex, reg_id_t *dest_reg O
      * probably not true if we fault the base).
      */
     opnd_size_t opsz = opnd_get_size(instr_get_src(strex, 0));
-    *dest_reg = reg_resize_to_opsz(opnd_get_reg(instr_get_dst(strex, 1)), opsz);
+    /* The store dest reg could equal a load dest reg, in which case the caller must
+     * pass us a scratch reg.
+     */
+    if (*dest_reg == DR_REG_NULL)
+        *dest_reg = reg_resize_to_opsz(opnd_get_reg(instr_get_dst(strex, 1)), opsz);
+    else
+        *dest_reg = reg_resize_to_opsz(*dest_reg, opsz);
```

Comment:
Instead of using the dest_reg if provided, shouldn't we check for the load-sharing-store-result case?

Reply:
No, see the comment on the line above: the caller has to pass us a scratch reg for that case.

Reply:
I mean: if it is the load-sharing-store-result case, assert that the caller provided a scratch reg in `dest_reg`; and if it is not that case, don't use the scratch even if it was provided by the user. What would happen currently in the load-sharing-store-result case if the caller didn't provide the scratch reg?

Reply:
This function does not know what the load looks like so it doesn't know whether it's the load-sharing-store-result case: it is passed only the store, and the caller does not know what the load is for slowpath cases.

> What would happen currently in the load-sharing-store-result case if the caller didn't provide the scratch reg?

Probably an infinite loop.

Reply:
So summarizing: an assert could only go in the caller for the optimized case; but that's where the caller is already picking the scratch reg so it wouldn't help.  Unfortunately I don't think more can be done.  This code is getting complex but simplifying by dropping some of these cases has its own downsides...I think we have to live with it.  A virtual scratch register layer could help but has its own problems.